### PR TITLE
feat(aws-android-sdk-cognitoidentityprovider): support custom endpoint

### PR DESCRIPTION
**Pull request by thisisabhash**
*Issue #, if available:* NA

*Description of changes:* Support for Custom endpoints in Cognito User Identity Pool

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Review by richardmcclellan — APPROVED** [review]
Overall looks good.  In the absence of unit tests (unfortunately it is very difficult to test `AWSMobileClient` given its current architecture), could you explain what manual testing you've done to confirm this change works?

File: aws-android-sdk-cognitoidentityprovider/src/main/java/com/amazonaws/mobileconnectors/cognitoidentityprovider/CognitoUserPool.java [hunk 1]
```diff
@@ -274,6 +274,26 @@ public CognitoUserPool(Context context, String userPoolId, String clientId, Stri
...
+     * @param context               REQUIRED: Android application context.
+     * @param userPoolId            REQUIRED: User-pool-Id of the user-pool.
+     * @param clientId              REQUIRED: Client-Id generated for this app and user-pool at the
+     *                              Cognito Identity Provider developer console.
+     * @param clientSecret          REQUIRED: Client Secret generated for this app and user-pool at
+     *                              the Cognito Identity Provider developer console.
+     * @param pinpointAppId         REQUIRED: AWS Pinpoint App Id for analytics.
+     * @param cognitoUserPoolCustomEndpoint REQUIRED: Custom endpoint for Cognito Userpool
+     */
+    public CognitoUserPool(Context context, String userPoolId, String clientId, String clientSecret, AmazonCognitoIdentityProvider client, String pinpointAppId, String cognitoUserPoolCustomEndpoint) {
```
Comment (jamesonwilliams) [inline aws-android-sdk-cognitoidentityprovider/src/main/java/com/amazonaws/mobileconnectors/cognitoidentityprovider/CognitoUserPool.java:289]: If this constructor has the largest number of arguments, it probably makes sense to have it implement all of the assignments itself.

The existing constructor (which does not have a `cognitoUserPoolCustomEndpoint`) would then call this with `null` for the newly added argument.

File: aws-android-sdk-cognitoidentityprovider/src/main/java/com/amazonaws/mobileconnectors/cognitoidentityprovider/CognitoUserPool.java [hunk 2]
```diff
@@ -589,4 +609,11 @@ protected UserContextDataType getUserContextData(String userId) {
         }
         return contextData;
     }
+
+    /**
+     * Returns the {@link AmazonCognitoIdentityProvider} instance
+     */
+    public AmazonCognitoIdentityProvider getClientInstance() {
```
Comment (jamesonwilliams) [inline aws-android-sdk-cognitoidentityprovider/src/main/java/com/amazonaws/mobileconnectors/cognitoidentityprovider/CognitoUserPool.java:616]: This codebase has multiple nested layers of accretion, and there are lots of different clients. It'd probably be better to be explicit, here:

```suggestion
    public AmazonCognitoIdentityProvider getAmazonCognitoIdentityProvider() {
```

File: aws-android-sdk-cognitoidentityprovider/src/test/java/CognitoIdentityProviderCustomEndPointTest.java [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+
+    /**
+     * Test that end point is set for the {@link AmazonCognitoIdentityProviderClient } with correct value
+     * when {@link CognitoUserPool#CognitoUserPool(Context, String, String, String, AmazonCognitoIdentityProvider, String, String)} is called
+     * with a valid endpoint
+     */
+    @Test
+    public void testForCustomEndpoint() {
+        testPool = new CognitoUserPool(appContext, TEST_USER_POOL, TEST_CLIENT_ID, TEST_CLIENT_SECRET, mockCustomEndpointClient, TEST_PINPOINT_APP_ID, DUMMY_CUSTOM_ENDPOINT);
+        verify(testPool.getClientInstance()).setEndpoint(DUMMY_CUSTOM_ENDPOINT);
```
Comment (jamesonwilliams) [inline aws-android-sdk-cognitoidentityprovider/src/test/java/CognitoIdentityProviderCustomEndPointTest.java:61]:
```suggestion
        verify(mockCustomEndpointClient).setEndpoint(DUMMY_CUSTOM_ENDPOINT);
```

File: aws-android-sdk-cognitoidentityprovider/src/test/java/CognitoIdentityProviderCustomEndPointTest.java [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+
+    /**
+     * Test that end point is not set for the {@link AmazonCognitoIdentityProviderClient }
+     * when {@link CognitoUserPool#CognitoUserPool(Context, String, String, String, AmazonCognitoIdentityProvider, String, String)} is called
+     * with a null endpoint
+     */
+    @Test
+    public void testForCustomEndpointWhenNull() {
+        testPool = new CognitoUserPool(appContext, TEST_USER_POOL, TEST_CLIENT_ID, TEST_CLIENT_SECRET, mockCustomEndpointClient, TEST_PINPOINT_APP_ID, null);
+        verify(testPool.getClientInstance(), never()).setEndpoint(null);
```
Comment (jamesonwilliams) [inline aws-android-sdk-cognitoidentityprovider/src/test/java/CognitoIdentityProviderCustomEndPointTest.java:72]: Strongest assertion would be `verifyNoInvocations(mockCustomEndpointClient)`.

Next strongest would be `verify(mockCustomEndpointClient, never()).setEndpoint(any())` (which doesn't match `null`.)

Current assertion just says there is never a `setEndpoint` call with `null` as the arg, which is weaker than the above assertions.

**Review by jamesonwilliams — APPROVED** [review]
Nitpicks below, looks good.

File: aws-android-sdk-cognitoidentityprovider/src/test/java/CognitoIdentityProviderCustomEndPointTest.java (hunk 4 above)
Comment (thisisabhash) [reply aws-android-sdk-cognitoidentityprovider/src/test/java/CognitoIdentityProviderCustomEndPointTest.java:72]: Did you mean verifyZeroInteractions(mockCustomEndpointClient) ?

**Review by jamesonwilliams — APPROVED** [review]

**Comment by thisisabhash** [comment]
Manual testing done with signUp(), confirmSignUp() and signIn() Auth APIs to check for custom endpoint.

**Pull request merged**